# Zastosowanie  Pythona w analizie danych w moim zawodzie 
### Poniżej opis propozycji wykożystania Pythona w zanalizie danych ziązanych z nadzorowaniem nad umową prania najmu i konserwacji odzieży roboczej

#### ID_98
#### Grupa:	1Z132-IS

Opis zagadnienia:
Nadzór nad umową zakłada wyposażenie pracowników Spółki w odzież której czasookres użytkowania wynosi 50 cykli prań przemysłowych. Wyposażenie zależy od grupy przypisania pracowników. Dane pobieramy z Systemu TETA (informacje o przypisania pracowników do grupy zaopatrzenia. Dane eksportowane w formacie csv) dane z systemów pralniczych firmy zewnętrznej MY_ELIS (dane eksportowane w formacie xmlx)

_Z uwagi na ograniczenia czasowe ograniczyłem się do przedstawienia kilku przykładowych czynności_


#### Pracę zaczynamy od importu pakietów
***

W module os znajduje się szereg funkcji umożliwiających pozyskiwanie informacji na temat plików i folderów znajdujących się na dysku twardym komputera a także tworzenie i usuwanie tychże plików.

Modół datetime będziemy wykożystywać do nadania nazwie pliku sygnaturę czasu. 


In [1]:
import pandas as pd
import numpy as np
import os
import datetime

Najpierw definiujemy ścieżkę z której pobierzemy pliki pobrane z systemu raportowego pralni za pomocą platformy MY_ELIS. Polityka bezpieczeństwa wykonawcy nie pozwala na bezośrednie pobranie danych z bazy danych.
Następnie tworzymy zmienną w której zapisujemy pliki znajdujące sięw katalogu.
Pliki mają sygnaturę zakładu pralniczego. Wykonawca posiada 6 pralni.

In [2]:
path = r'F:\ELIS'      # I'm sorry but I don't share data files

In [3]:
lista_plikow = os.listdir(path)

In [4]:
len(lista_plikow)

6

In [5]:
print(lista_plikow)

['GarmentReport-POZNAŃ.xlsx', 'GarmentReport-SIEMIANOWICE.xlsx', 'GarmentReport-TARNÓW.xlsx', 'GarmentReport-WARSZAWA.xlsx', 'GarmentReport-WROCŁAW.xlsx', 'GarmentReport-ZUKOWO.xlsx']


#### Wczytujemy przykąłdowy plik aby wstępnie zapoznać sięz danymi

In [6]:
db_0 = pd.read_excel(r'F:\ELIS\GarmentReport-POZNAŃ.xlsx')

In [7]:
db_0.head()

nr adresu dostawy                          Nazwa Tożsamość odzieży  \
0             708751  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2717   
1             708751  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2FD0   
2             708751  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32B7   
3             708751  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32ED   
4             708751  GAZ - SYSTEM O/ POZNAŃ GROBLA       6C0001B13AC   

   Nr artykułu                                    Opis Rozmiar  \
0       101045  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO      50   
1       101045  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO      NS   
2       101045  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO      NS   
3       101045  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO      NS   
4       101045  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO      52   

  Ostatnie pranie      Wkład  Ilość prań                   Typ  ... Nr konta  \
0      2022-08-30 2022-08-30           1  ARTYKUŁY WYNAJMOWANE  ...  P130552   
1      2021-01-05 2021-01-05           0  ARTYKUŁY WYNAJMOWANE  ...  C181459   
2      2021-01-05 2021-01-05           0  ARTYKUŁY WYNAJMOWANE  ...  C181459   
3      2020-12-30 2020-12-30           0  ARTYKUŁY WYNAJMOWANE  ...  P130252   
4      2022-11-16 2022-11-15           1  ARTYKUŁY WYNAJMOWANE  ...  P130552   

  Nr zatrud.      Imię Nazwisko Dział  Szafka  Skrytka  \
0         82  P0000751      NaN   NaN   512.0      3.0   
1          3  C0000765      NaN   3.0   500.0      3.0   
2          3  C0000765      NaN   3.0   500.0      3.0   
3         26  P0000067      NaN   5.0   505.0      1.0   
4         87  P0000758      NaN   NaN   512.0      5.0   

   Liczba prań dla aktualnego użytkownika odzieży.  Zakończył  \
0                                                0        NaN   
1                                                0        NaN   
2                                                0        NaN   
3                                                0        NaN   
4                                                1        NaN   

  Wartość resztkowa  
0            234.22  
1            239.00  
2            239.00  
3            234.22  
4            654.86  

[5 rows x 21 columns]

In [8]:
db_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9382 entries, 0 to 9381
Data columns (total 21 columns):
 #   Column                                           Non-Null Count  Dtype         
---  ------                                           --------------  -----         
 0   nr adresu dostawy                                9382 non-null   int64         
 1   Nazwa                                            9382 non-null   object        
 2   Tożsamość odzieży                                9382 non-null   object        
 3   Nr artykułu                                      9382 non-null   int64         
 4   Opis                                             9382 non-null   object        
 5   Rozmiar                                          9382 non-null   object        
 6   Ostatnie pranie                                  9382 non-null   datetime64[ns]
 7   Wkład                                            9382 non-null   datetime64[ns]
 8   Ilość prań                            

In [9]:
db_0.columns

Index(['nr adresu dostawy', 'Nazwa', 'Tożsamość odzieży', 'Nr artykułu',
       'Opis', 'Rozmiar', 'Ostatnie pranie', 'Wkład', 'Ilość prań', 'Typ',
       'Oznakowania', 'Nr konta', 'Nr zatrud.', 'Imię', 'Nazwisko', 'Dział',
       'Szafka', 'Skrytka', 'Liczba prań dla aktualnego użytkownika odzieży.',
       'Zakończył', 'Wartość resztkowa'],
      dtype='object')

Do dalszej analizy (prykładu) nie będziemy potrzebować wszystkich kolumn. Przegląd pliku był potrzebny do wybrania danych niezbędnych do dalszej analizy.

w tym przypadku wybrano: 

    *'Nazwa' - *nazwa jednostki do której jest przypisana odzież* 
    *'Tożsamość odzieży' - *unikalna nazwa konkretnej sztuki odzieży*
    *'Opis' - *opis typu odzież *
    *'Ilość prań' - *ilość prań na danej sztuce odzieży*
    *'Typ' - *opisuje własność odzieży*
    *'Imię','Liczba prań dla aktualnego użytkownika odzieży.'



In [10]:
number =1

### Wczytujemy pliki z katalogu 
***
wykożystujemy do tego pętle *for* w celu automatycznego utworzenia zmiennej używamy funkcji globals()


In [11]:
for i in lista_plikow:
    a = os.path.join(path,i)
    globals()['db_{}'.format(number)] = pd.read_excel(a, usecols=['Nazwa','Tożsamość odzieży','Opis', 'Ilość prań', 'Typ','Imię','Liczba prań dla aktualnego użytkownika odzieży.' ])
    number +=1


Możemy podglądnąć przykładowe dataframe które powstały z wczytania plików


In [12]:
db_1.head()

Nazwa Tożsamość odzieży  \
0  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2717   
1  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2FD0   
2  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32B7   
3  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32ED   
4  GAZ - SYSTEM O/ POZNAŃ GROBLA       6C0001B13AC   

                                     Opis  Ilość prań                   Typ  \
0  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           1  ARTYKUŁY WYNAJMOWANE   
1  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   
2  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   
3  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   
4  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           1  ARTYKUŁY WYNAJMOWANE   

       Imię  Liczba prań dla aktualnego użytkownika odzieży.  
0  P0000751                                                0  
1  C0000765                                                0  
2  C0000765                                                0  
3  P0000067                                                0  
4  P0000758                                                1

In [13]:
db_2.head()

Nazwa Tożsamość odzieży  \
0  GAZ-SYSTEM O/ŚWIERKLANY BIELSKO       40000F9B71A   
1  GAZ-SYSTEM O/ŚWIERKLANY BIELSKO       40000FA15C4   
2  GAZ-SYSTEM O/ŚWIERKLANY BIELSKO       40000FA2595   
3  GAZ-SYSTEM O/ŚWIERKLANY BIELSKO       40000FA25F7   
4  GAZ-SYSTEM O/ŚWIERKLANY BIELSKO       40000FA2627   

                                     Opis  Ilość prań                   Typ  \
0  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           1  ARTYKUŁY WYNAJMOWANE   
1  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           5  ARTYKUŁY WYNAJMOWANE   
2  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   
3  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   
4  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   

       Imię  Liczba prań dla aktualnego użytkownika odzieży.  
0  S0408788                                                1  
1  S0456412                                                5  
2  S0466638                                                0  
3  S0466638                                                0  
4  S0466638                                                0

In [14]:
db_1.shape

(9382, 7)

In [15]:
db_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9382 entries, 0 to 9381
Data columns (total 7 columns):
 #   Column                                           Non-Null Count  Dtype 
---  ------                                           --------------  ----- 
 0   Nazwa                                            9382 non-null   object
 1   Tożsamość odzieży                                9382 non-null   object
 2   Opis                                             9382 non-null   object
 3   Ilość prań                                       9382 non-null   int64 
 4   Typ                                              9382 non-null   object
 5   Imię                                             9381 non-null   object
 6   Liczba prań dla aktualnego użytkownika odzieży.  9382 non-null   int64 
dtypes: int64(2), object(5)
memory usage: 513.2+ KB


Łączymy wczytane dane w jedną dataframe

In [16]:
df = pd.concat(objs=(db_1, db_2, db_3, db_4, db_5 ,db_6), ignore_index=True)

In [17]:
df.head()

Nazwa Tożsamość odzieży  \
0  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2717   
1  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2FD0   
2  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32B7   
3  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32ED   
4  GAZ - SYSTEM O/ POZNAŃ GROBLA       6C0001B13AC   

                                     Opis  Ilość prań                   Typ  \
0  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           1  ARTYKUŁY WYNAJMOWANE   
1  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   
2  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   
3  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  ARTYKUŁY WYNAJMOWANE   
4  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           1  ARTYKUŁY WYNAJMOWANE   

       Imię  Liczba prań dla aktualnego użytkownika odzieży.  
0  P0000751                                                0  
1  C0000765                                                0  
2  C0000765                                                0  
3  P0000067                                                0  
4  P0000758                                                1

In [19]:
df.shape

(34691, 7)

Proszęzwrócić uwagę na ilość wierszy w dataframe db_1 miało 9382 po połączeniu df ma 34691

In [20]:
df['Typ'].unique()

array(['ARTYKUŁY WYNAJMOWANE', 'WŁASNOŚĆ KLIENTA'], dtype=object)

In [21]:
df['Opis'].unique()

array(['BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO',
       'BLUZY - WŁASNOŚĆ KLIENTA GAZ SYSTEM',
       'KAMIZELKA - WŁASNOŚĆ KLIENTA GAZ SYSTEM',
       'KURTKA OC. M GRANAT-ŻÓŁT MULTI HIVI LOGO',
       'KURTKI - WŁASNOŚĆ KLIENTA GAZ SYSTEM',
       'OGROD.OCIEPL-WŁASNOŚĆ KLIENTA GAZ SYSTEM',
       'OGRODNICZKI-WŁASNOŚĆ KLIENTA GAZ SYSTEM',
       'POLO - WŁASNOŚĆ KLIENTA GAZ SYSTEM',
       'POLO DŁ.RĘK M GRANAT-ŻÓŁ MULTI HIVI LOGO',
       'SOFTSHELL GRANAT-ŻÓŁTY HIVI MULTIO  LOGO',
       'SPODNIE - WŁASNOŚĆ KLIENTA GAZ SYSTEM',
       'SPODNIE M GRANAT-ŻÓŁTY MULTIOCH HIVI',
       'KAMIZELKA OC  M GRAN-ŻÓŁ MULTI HIVI LOGO',
       'OGRODNICZ OC. M GRAN-ŻÓŁ MULTI HIVI LOGO',
       'OGRODNICZK M GRANAT-ŻÓŁT MULTI HIVI LOGO'], dtype=object)

W naszej analizie nie interesują nas artykuły stanowiące 'WŁASNOŚĆ KLIENTA' dlatego też odfiltrowuje niepotrzbne dane

In [22]:
df = df[df['Typ'] == 'ARTYKUŁY WYNAJMOWANE']

In [23]:
df.shape

(32690, 7)

In [24]:
df.drop('Typ', axis=1, inplace=True)

In [25]:
df.head()

Nazwa Tożsamość odzieży  \
0  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2717   
1  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2FD0   
2  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32B7   
3  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32ED   
4  GAZ - SYSTEM O/ POZNAŃ GROBLA       6C0001B13AC   

                                     Opis  Ilość prań      Imię  \
0  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           1  P0000751   
1  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  C0000765   
2  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  C0000765   
3  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           0  P0000067   
4  BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO           1  P0000758   

   Liczba prań dla aktualnego użytkownika odzieży.  
0                                                0  
1                                                0  
2                                                0  
3                                                0  
4                                                1

In [26]:
df['Opis'].unique()

array(['BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO',
       'KURTKA OC. M GRANAT-ŻÓŁT MULTI HIVI LOGO',
       'POLO DŁ.RĘK M GRANAT-ŻÓŁ MULTI HIVI LOGO',
       'SOFTSHELL GRANAT-ŻÓŁTY HIVI MULTIO  LOGO',
       'SPODNIE M GRANAT-ŻÓŁTY MULTIOCH HIVI',
       'KAMIZELKA OC  M GRAN-ŻÓŁ MULTI HIVI LOGO',
       'OGRODNICZ OC. M GRAN-ŻÓŁ MULTI HIVI LOGO',
       'OGRODNICZK M GRANAT-ŻÓŁT MULTI HIVI LOGO'], dtype=object)

Patrząc na opis niekoniecznie chcemy mieć takie dane dlatego zmienimy je na coś co się będzie łatwiej czytać.
W tym celu przygotowujemy słownik a następnie dokonamy zmiany opisów

In [27]:
dict_1 = {'BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO':'BLUZA GRANAT','KURTKA OC. M GRANAT-ŻÓŁT MULTI HIVI LOGO' : "KURTKA OC.",'POLO DŁ.RĘK M GRANAT-ŻÓŁ MULTI HIVI LOGO':'POLO', 'SOFTSHELL GRANAT-ŻÓŁTY HIVI MULTIO  LOGO' :'SOFTSHELL',
         'SPODNIE M GRANAT-ŻÓŁTY MULTIOCH HIVI': 'SPODNIE', 'KAMIZELKA OC  M GRAN-ŻÓŁ MULTI HIVI LOGO' : 'KAMIZELKA OC',
         'OGRODNICZ OC. M GRAN-ŻÓŁ MULTI HIVI LOGO':'OGRODNICZ OC.', 'OGRODNICZK M GRANAT-ŻÓŁT MULTI HIVI LOGO': 'OGRODNICZKI'}

In [28]:
dict_1

{'BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO': 'BLUZA GRANAT',
 'KURTKA OC. M GRANAT-ŻÓŁT MULTI HIVI LOGO': 'KURTKA OC.',
 'POLO DŁ.RĘK M GRANAT-ŻÓŁ MULTI HIVI LOGO': 'POLO',
 'SOFTSHELL GRANAT-ŻÓŁTY HIVI MULTIO  LOGO': 'SOFTSHELL',
 'SPODNIE M GRANAT-ŻÓŁTY MULTIOCH HIVI': 'SPODNIE',
 'KAMIZELKA OC  M GRAN-ŻÓŁ MULTI HIVI LOGO': 'KAMIZELKA OC',
 'OGRODNICZ OC. M GRAN-ŻÓŁ MULTI HIVI LOGO': 'OGRODNICZ OC.',
 'OGRODNICZK M GRANAT-ŻÓŁT MULTI HIVI LOGO': 'OGRODNICZKI'}

In [29]:
list(dict_1.keys())[0]

'BLUZA GRANAT-ŻÓŁTY HIVI MULTIOCHR LOGO'

In [30]:
list(dict_1.values())[0]

'BLUZA GRANAT'

In [31]:
for key in list(dict_1.keys()):
    df.loc[df['Opis'] == key, 'Opis'] = dict_1[key] 

In [32]:
df.head()

Nazwa Tożsamość odzieży          Opis  Ilość prań  \
0  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2717  BLUZA GRANAT           1   
1  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA2FD0  BLUZA GRANAT           0   
2  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32B7  BLUZA GRANAT           0   
3  GAZ - SYSTEM O/ POZNAŃ GROBLA       40000FA32ED  BLUZA GRANAT           0   
4  GAZ - SYSTEM O/ POZNAŃ GROBLA       6C0001B13AC  BLUZA GRANAT           1   

       Imię  Liczba prań dla aktualnego użytkownika odzieży.  
0  P0000751                                                0  
1  C0000765                                                0  
2  C0000765                                                0  
3  P0000067                                                0  
4  P0000758                                                1

In [33]:
df['Opis'].unique()

array(['BLUZA GRANAT', 'KURTKA OC.', 'POLO', 'SOFTSHELL', 'SPODNIE',
       'KAMIZELKA OC', 'OGRODNICZ OC.', 'OGRODNICZKI'], dtype=object)

### Przyglądniemy się teraz bardziej co mamy  w kolumnie 'Nazwa'
***
W kolumnie widać że dane są różnie wprowadzane w zależności od pralni, ponadto mamy informacje których na danym etapie analizy nie potrzebujemy takie jak np magazyn buforowy lub elementy wstrzymane do wyjaśneinia. 
Poniżej przykładowo ograniczymy dane

In [34]:
df['Nazwa'].unique()

array(['GAZ - SYSTEM O/ POZNAŃ GROBLA',
       'GAZ - SYSTEM O/ POZNAŃ CZERWONACKA 5',
       'GAZ - SYSTEM O/POZNAŃ,  PIŁA', 'GAZ - SYSTEM O/POZNAŃ,  GRANOWO',
       'GAZ - SYSTEM O/POZNAŃ,  KROBIA', 'GAZ - SYSTEM O/POZNAŃ,  LWÓWEK',
       'GAZ - SYSTEM O/ POZNAŃ, ODOLANÓW',
       'GAZ - SYSTEM O/POZNAŃ,  KALISZ', 'GAZ - SYSTEM O/ POZNAŃ, LESZNO',
       'GAZ - SYSTEM O/ POZNAŃ, SZCZECIN',
       'GAZ - SYSTEM O/POZNAŃ,  GOLENIÓW',
       'GAZ - SYSTEM O/ POZNAŃ, GORZÓW WIELKOPOL',
       'GAZ - SYSTEM O/POZNAŃ,  UJŚCIE',
       'GAZ - SYSTEM O/WROCŁAW, ZIELONA GÓRA',
       'GAZ-SYSTEM / MAGAZYN BUFOR',
       'GAZ - SYSTEM O/POZNAŃ,  ŚWINOUJŚCIE',
       'GAZ SYSTEM\\ DO WYJAŚNIENIA', 'GAZ - SYSTEM O/ POZNAŃ, NIECHORZE',
       'GAZ-SYSTEM O/ŚWIERKLANY BIELSKO',
       'GAZ-SYSTEM O/ŚWIERKLANY CZĘSTOCHOWA',
       'GAZ-SYSTEM O/TARNÓW KRAKÓW', 'GAZ-SYSTEM O/ŚWIERKLANY OPOLE',
       'GAZ-SYSTEM O/ŚWIERKLANY RUDA ŚLĄSKA',
       'GAZ-SYSTEM O/ŚWIERKLANY ŚWIERKLANY',
       'GAZ SY

#### Na dalszym etapie nie będziemy potrzebować informacji dotyczących odzieży znajdujących się na magazynie buforowym ani tymczasowo przekazanych do wyjaśnienia

Do pozbycia siętych danych utworzymy filtr przy użyciu metody str **contains** 

In [35]:
'GDAŃSK' in 'GAZ SYSTEM O/GDAŃSK- GRUDZIĄDZ'

True

In [36]:
Filtr_1 = ~df['Nazwa'].str.contains('BUFOR')

In [37]:
df = df[Filtr_1]

In [38]:
df['Nazwa'].unique()

array(['GAZ - SYSTEM O/ POZNAŃ GROBLA',
       'GAZ - SYSTEM O/ POZNAŃ CZERWONACKA 5',
       'GAZ - SYSTEM O/POZNAŃ,  PIŁA', 'GAZ - SYSTEM O/POZNAŃ,  GRANOWO',
       'GAZ - SYSTEM O/POZNAŃ,  KROBIA', 'GAZ - SYSTEM O/POZNAŃ,  LWÓWEK',
       'GAZ - SYSTEM O/ POZNAŃ, ODOLANÓW',
       'GAZ - SYSTEM O/POZNAŃ,  KALISZ', 'GAZ - SYSTEM O/ POZNAŃ, LESZNO',
       'GAZ - SYSTEM O/ POZNAŃ, SZCZECIN',
       'GAZ - SYSTEM O/POZNAŃ,  GOLENIÓW',
       'GAZ - SYSTEM O/ POZNAŃ, GORZÓW WIELKOPOL',
       'GAZ - SYSTEM O/POZNAŃ,  UJŚCIE',
       'GAZ - SYSTEM O/WROCŁAW, ZIELONA GÓRA',
       'GAZ - SYSTEM O/POZNAŃ,  ŚWINOUJŚCIE',
       'GAZ SYSTEM\\ DO WYJAŚNIENIA', 'GAZ - SYSTEM O/ POZNAŃ, NIECHORZE',
       'GAZ-SYSTEM O/ŚWIERKLANY BIELSKO',
       'GAZ-SYSTEM O/ŚWIERKLANY CZĘSTOCHOWA',
       'GAZ-SYSTEM O/TARNÓW KRAKÓW', 'GAZ-SYSTEM O/ŚWIERKLANY OPOLE',
       'GAZ-SYSTEM O/ŚWIERKLANY RUDA ŚLĄSKA',
       'GAZ-SYSTEM O/ŚWIERKLANY ŚWIERKLANY',
       'GAZ SYSTEM O/ŚWIERKLANY KĘDZIERZYN KOŹLE',


In [39]:
Filtr_2 = ~df['Nazwa'].str.contains('WYJAŚNIENIA')

In [40]:
df = df[Filtr_2]

In [41]:
df['Nazwa'].unique()

array(['GAZ - SYSTEM O/ POZNAŃ GROBLA',
       'GAZ - SYSTEM O/ POZNAŃ CZERWONACKA 5',
       'GAZ - SYSTEM O/POZNAŃ,  PIŁA', 'GAZ - SYSTEM O/POZNAŃ,  GRANOWO',
       'GAZ - SYSTEM O/POZNAŃ,  KROBIA', 'GAZ - SYSTEM O/POZNAŃ,  LWÓWEK',
       'GAZ - SYSTEM O/ POZNAŃ, ODOLANÓW',
       'GAZ - SYSTEM O/POZNAŃ,  KALISZ', 'GAZ - SYSTEM O/ POZNAŃ, LESZNO',
       'GAZ - SYSTEM O/ POZNAŃ, SZCZECIN',
       'GAZ - SYSTEM O/POZNAŃ,  GOLENIÓW',
       'GAZ - SYSTEM O/ POZNAŃ, GORZÓW WIELKOPOL',
       'GAZ - SYSTEM O/POZNAŃ,  UJŚCIE',
       'GAZ - SYSTEM O/WROCŁAW, ZIELONA GÓRA',
       'GAZ - SYSTEM O/POZNAŃ,  ŚWINOUJŚCIE',
       'GAZ - SYSTEM O/ POZNAŃ, NIECHORZE',
       'GAZ-SYSTEM O/ŚWIERKLANY BIELSKO',
       'GAZ-SYSTEM O/ŚWIERKLANY CZĘSTOCHOWA',
       'GAZ-SYSTEM O/TARNÓW KRAKÓW', 'GAZ-SYSTEM O/ŚWIERKLANY OPOLE',
       'GAZ-SYSTEM O/ŚWIERKLANY RUDA ŚLĄSKA',
       'GAZ-SYSTEM O/ŚWIERKLANY ŚWIERKLANY',
       'GAZ SYSTEM O/ŚWIERKLANY KĘDZIERZYN KOŹLE',
       'GAZ-SYSTEM O/TARNÓW- JA

In [42]:
df['Nazwa'].str.split('/').str[1].unique()

array([' POZNAŃ GROBLA', ' POZNAŃ CZERWONACKA 5', 'POZNAŃ,  PIŁA',
       'POZNAŃ,  GRANOWO', 'POZNAŃ,  KROBIA', 'POZNAŃ,  LWÓWEK',
       ' POZNAŃ, ODOLANÓW', 'POZNAŃ,  KALISZ', ' POZNAŃ, LESZNO',
       ' POZNAŃ, SZCZECIN', 'POZNAŃ,  GOLENIÓW',
       ' POZNAŃ, GORZÓW WIELKOPOL', 'POZNAŃ,  UJŚCIE',
       'WROCŁAW, ZIELONA GÓRA', 'POZNAŃ,  ŚWINOUJŚCIE',
       ' POZNAŃ, NIECHORZE', 'ŚWIERKLANY BIELSKO',
       'ŚWIERKLANY CZĘSTOCHOWA', 'TARNÓW KRAKÓW', 'ŚWIERKLANY OPOLE',
       'ŚWIERKLANY RUDA ŚLĄSKA', 'ŚWIERKLANY ŚWIERKLANY',
       'ŚWIERKLANY KĘDZIERZYN KOŹLE', 'TARNÓW- JAROSŁAW',
       'TARNÓW LUBACZÓW', 'TARNÓW MAĆKOWICE', 'TARNÓW POGÓRSKA WOLA',
       'TARNÓW- RZESZÓW', 'TARNÓW- SANDOMIERZ', 'TARNÓW STRACHOCINA',
       'TARNÓW JASŁO - STRACHOCINA', ' TARNÓW CIESZACIN MAŁY',
       'W-WA, SARNAKI KLEPACZEW', 'W-WA, MSZCZONOWSKA',
       'W-WA, NIEPORĘT, JANA KAZIEM', 'W-WA,KOŃSKOWOLA, WARSZAWSKA',
       'W-WA, SUWAŁKI, INNOWACYJNA', 'W-WA, ZGIERZ, PIĄTKOWSKA',
       'TARN

In [43]:
df['Nazwa'] = df['Nazwa'].str.split('/').str[1]

In [44]:
df['Nazwa'] = df['Nazwa'].str.replace('-','')

In [45]:
df['Nazwa'] = df['Nazwa'].str.replace(',',' ')

In [46]:
df['Nazwa'] = df['Nazwa'].str.replace('  ',' ')

In [47]:
df['Nazwa'] = df['Nazwa'].str.replace('  ',' ')

In [48]:
df['Nazwa'] = df['Nazwa'].str.strip()

In [49]:
df['Nazwa'] = df['Nazwa'].str.replace('WWA','REMBELSZCZYZNA')

In [50]:
df['Nazwa'] = df['Nazwa'].str.replace('REMBELSZCZYZNA MSZCZONOWSKA','CENTRALA MSZCZONOWSKA')

In [51]:
df['Nazwa'].unique()

array(['POZNAŃ GROBLA', 'POZNAŃ CZERWONACKA 5', 'POZNAŃ PIŁA',
       'POZNAŃ GRANOWO', 'POZNAŃ KROBIA', 'POZNAŃ LWÓWEK',
       'POZNAŃ ODOLANÓW', 'POZNAŃ KALISZ', 'POZNAŃ LESZNO',
       'POZNAŃ SZCZECIN', 'POZNAŃ GOLENIÓW', 'POZNAŃ GORZÓW WIELKOPOL',
       'POZNAŃ UJŚCIE', 'WROCŁAW ZIELONA GÓRA', 'POZNAŃ ŚWINOUJŚCIE',
       'POZNAŃ NIECHORZE', 'ŚWIERKLANY BIELSKO', 'ŚWIERKLANY CZĘSTOCHOWA',
       'TARNÓW KRAKÓW', 'ŚWIERKLANY OPOLE', 'ŚWIERKLANY RUDA ŚLĄSKA',
       'ŚWIERKLANY ŚWIERKLANY', 'ŚWIERKLANY KĘDZIERZYN KOŹLE',
       'TARNÓW JAROSŁAW', 'TARNÓW LUBACZÓW', 'TARNÓW MAĆKOWICE',
       'TARNÓW POGÓRSKA WOLA', 'TARNÓW RZESZÓW', 'TARNÓW SANDOMIERZ',
       'TARNÓW STRACHOCINA', 'TARNÓW JASŁO STRACHOCINA',
       'TARNÓW CIESZACIN MAŁY', 'REMBELSZCZYZNA SARNAKI KLEPACZEW',
       'CENTRALA MSZCZONOWSKA', 'REMBELSZCZYZNA NIEPORĘT JANA KAZIEM',
       'REMBELSZCZYZNA KOŃSKOWOLA WARSZAWSKA',
       'REMBELSZCZYZNA SUWAŁKI INNOWACYJNA',
       'REMBELSZCZYZNA ZGIERZ PIĄTKOWSKA', 'T

In [52]:
df.head()

Nazwa Tożsamość odzieży          Opis  Ilość prań      Imię  \
0  POZNAŃ GROBLA       40000FA2717  BLUZA GRANAT           1  P0000751   
1  POZNAŃ GROBLA       40000FA2FD0  BLUZA GRANAT           0  C0000765   
2  POZNAŃ GROBLA       40000FA32B7  BLUZA GRANAT           0  C0000765   
3  POZNAŃ GROBLA       40000FA32ED  BLUZA GRANAT           0  P0000067   
4  POZNAŃ GROBLA       6C0001B13AC  BLUZA GRANAT           1  P0000758   

   Liczba prań dla aktualnego użytkownika odzieży.  
0                                                0  
1                                                0  
2                                                0  
3                                                0  
4                                                1

In [53]:
df[['Odział', 'Jednostka']] = df['Nazwa'].str.split(" ", expand=True, n=1)

In [54]:
df.head()

Nazwa Tożsamość odzieży          Opis  Ilość prań      Imię  \
0  POZNAŃ GROBLA       40000FA2717  BLUZA GRANAT           1  P0000751   
1  POZNAŃ GROBLA       40000FA2FD0  BLUZA GRANAT           0  C0000765   
2  POZNAŃ GROBLA       40000FA32B7  BLUZA GRANAT           0  C0000765   
3  POZNAŃ GROBLA       40000FA32ED  BLUZA GRANAT           0  P0000067   
4  POZNAŃ GROBLA       6C0001B13AC  BLUZA GRANAT           1  P0000758   

   Liczba prań dla aktualnego użytkownika odzieży.  Odział Jednostka  
0                                                0  POZNAŃ    GROBLA  
1                                                0  POZNAŃ    GROBLA  
2                                                0  POZNAŃ    GROBLA  
3                                                0  POZNAŃ    GROBLA  
4                                                1  POZNAŃ    GROBLA

#### W celu podglądu np jaki stan mamy w poszczególnych lokalizacji możemy użyć zapytań znanych np s języka SQL

In [56]:
df.query('Nazwa == "POZNAŃ GROBLA" and Opis == "BLUZA GRANAT" ')

Nazwa Tożsamość odzieży          Opis  Ilość prań      Imię  \
0    POZNAŃ GROBLA       40000FA2717  BLUZA GRANAT           1  P0000751   
1    POZNAŃ GROBLA       40000FA2FD0  BLUZA GRANAT           0  C0000765   
2    POZNAŃ GROBLA       40000FA32B7  BLUZA GRANAT           0  C0000765   
3    POZNAŃ GROBLA       40000FA32ED  BLUZA GRANAT           0  P0000067   
4    POZNAŃ GROBLA       6C0001B13AC  BLUZA GRANAT           1  P0000758   
..             ...               ...           ...         ...       ...   
124  POZNAŃ GROBLA       7C002A0B592  BLUZA GRANAT           0  P0000616   
125  POZNAŃ GROBLA       7C002A0B61A  BLUZA GRANAT           0  P0000090   
126  POZNAŃ GROBLA       7C002A0B620  BLUZA GRANAT           0  P0000638   
127  POZNAŃ GROBLA       7C002A0B63D  BLUZA GRANAT           0  P0000027   
128  POZNAŃ GROBLA       7C002A0B670  BLUZA GRANAT           0  P0000616   

     Liczba prań dla aktualnego użytkownika odzieży.  Odział Jednostka  
0                                                  0  POZNAŃ    GROBLA  
1                                                  0  POZNAŃ    GROBLA  
2                                                  0  POZNAŃ    GROBLA  
3                                                  0  POZNAŃ    GROBLA  
4                                                  1  POZNAŃ    GROBLA  
..                                               ...     ...       ...  
124                                                0  POZNAŃ    GROBLA  
125                                                0  POZNAŃ    GROBLA  
126                                                0  POZNAŃ    GROBLA  
127                                                0  POZNAŃ    GROBLA  
128                                                0  POZNAŃ    GROBLA  

[129 rows x 8 columns]

#### Możemy zidentyfikować największego czyścioszka

In [57]:
df['Liczba prań dla aktualnego użytkownika odzieży.'].max()

36

In [58]:
df[df['Liczba prań dla aktualnego użytkownika odzieży.'] == df['Liczba prań dla aktualnego użytkownika odzieży.'].max()]

Nazwa Tożsamość odzieży  Opis  Ilość prań      Imię  \
7050  POZNAŃ ŚWINOUJŚCIE       6C000736520  POLO          36  C0001748   

      Liczba prań dla aktualnego użytkownika odzieży.  Odział    Jednostka  
7050                                               36  POZNAŃ  ŚWINOUJŚCIE

In [59]:
df[df["Imię"] == 'C0001748']

Nazwa Tożsamość odzieży           Opis  Ilość prań  \
5722  POZNAŃ ŚWINOUJŚCIE       6C0001B1415   BLUZA GRANAT           0   
5780  POZNAŃ ŚWINOUJŚCIE       6C0001B1594   BLUZA GRANAT           0   
6041  POZNAŃ ŚWINOUJŚCIE       6C0002CA0B2   KAMIZELKA OC           0   
6267  POZNAŃ ŚWINOUJŚCIE       6C00073593A     KURTKA OC.           0   
6305  POZNAŃ ŚWINOUJŚCIE       6C0007429FC     KURTKA OC.           0   
6400  POZNAŃ ŚWINOUJŚCIE       40000FA2389  OGRODNICZ OC.           0   
6425  POZNAŃ ŚWINOUJŚCIE       40000FA24E0  OGRODNICZ OC.           0   
6610  POZNAŃ ŚWINOUJŚCIE       6C0001B143C    OGRODNICZKI           2   
6612  POZNAŃ ŚWINOUJŚCIE       6C0001B1444    OGRODNICZKI           2   
7050  POZNAŃ ŚWINOUJŚCIE       6C000736520           POLO          36   
7110  POZNAŃ ŚWINOUJŚCIE       6C0007365CB           POLO          35   
7390  POZNAŃ ŚWINOUJŚCIE       6C0007428B2      SOFTSHELL           0   
7412  POZNAŃ ŚWINOUJŚCIE       6C00074290E      SOFTSHELL           1   
7554  POZNAŃ ŚWINOUJŚCIE       6C0001B1510        SPODNIE           0   
7574  POZNAŃ ŚWINOUJŚCIE       6C0003AC629        SPODNIE           0   

          Imię  Liczba prań dla aktualnego użytkownika odzieży.  Odział  \
5722  C0001748                                                0  POZNAŃ   
5780  C0001748                                                0  POZNAŃ   
6041  C0001748                                                0  POZNAŃ   
6267  C0001748                                                0  POZNAŃ   
6305  C0001748                                                0  POZNAŃ   
6400  C0001748                                                0  POZNAŃ   
6425  C0001748                                                0  POZNAŃ   
6610  C0001748                                                2  POZNAŃ   
6612  C0001748                                                2  POZNAŃ   
7050  C0001748                                               36  POZNAŃ   
7110  C0001748                                               35  POZNAŃ   
7390  C0001748                                                0  POZNAŃ   
7412  C0001748                                                1  POZNAŃ   
7554  C0001748                                                0  POZNAŃ   
7574  C0001748                                                0  POZNAŃ   

        Jednostka  
5722  ŚWINOUJŚCIE  
5780  ŚWINOUJŚCIE  
6041  ŚWINOUJŚCIE  
6267  ŚWINOUJŚCIE  
6305  ŚWINOUJŚCIE  
6400  ŚWINOUJŚCIE  
6425  ŚWINOUJŚCIE  
6610  ŚWINOUJŚCIE  
6612  ŚWINOUJŚCIE  
7050  ŚWINOUJŚCIE  
7110  ŚWINOUJŚCIE  
7390  ŚWINOUJŚCIE  
7412  ŚWINOUJŚCIE  
7554  ŚWINOUJŚCIE  
7574  ŚWINOUJŚCIE

In [60]:
asortyment = df.groupby(by='Opis', axis=0)

In [61]:
asortyment

In [62]:
asortyment.size()

Opis
BLUZA GRANAT     6679
KAMIZELKA OC     2010
KURTKA OC.       3687
OGRODNICZ OC.    2022
OGRODNICZKI      4013
POLO             6696
SOFTSHELL        3694
SPODNIE          2716
dtype: int64

In [63]:
prania = df.groupby( by=['Ilość prań', 'Opis' ])

In [64]:
prania.size()

Ilość prań  Opis         
0           BLUZA GRANAT     3410
            KAMIZELKA OC     1201
            KURTKA OC.       2216
            OGRODNICZ OC.    1535
            OGRODNICZKI       672
                             ... 
28          POLO                3
            SPODNIE             1
32          OGRODNICZKI         1
35          POLO                1
36          POLO                1
Length: 162, dtype: int64

In [65]:
prania.get_group((28,'POLO'))

Nazwa Tożsamość odzieży  Opis  Ilość prań      Imię  \
2883   POZNAŃ ODOLANÓW       7C001AFE41A  POLO          28  P0000526   
2885   POZNAŃ ODOLANÓW       7C001AFE426  POLO          28  P0000526   
15984  TARNÓW JAROSŁAW       7C001AFB707  POLO          28  T0030158   

       Liczba prań dla aktualnego użytkownika odzieży.  Odział Jednostka  
2883                                                28  POZNAŃ  ODOLANÓW  
2885                                                28  POZNAŃ  ODOLANÓW  
15984                                               28  TARNÓW  JAROSŁAW

#### Możemy przygotować tabelę przestawną

In [66]:
df.pivot_table(index = "Opis", columns='Ilość prań', aggfunc='count', fill_value=0, margins=True)

Imię                                                    ...  \
Ilość prań         0     1     2     3     4     5    6    7    8    9  ...   
Opis                                                                    ...   
BLUZA GRANAT    3410  1046   566   394   279   203  184  142  103   92  ...   
KAMIZELKA OC    1201   365   175   119    56    36   25   12   10    5  ...   
KURTKA OC.      2216   723   365   175    98    44   34   15    9    1  ...   
OGRODNICZ OC.   1535   279    98    45    35    21    5    2    1    1  ...   
OGRODNICZKI      672   236   204   204   188   207  210  216  213  223  ...   
POLO            1782   636   517   436   405   350  378  335  336  228  ...   
SOFTSHELL       2019   647   405   247   153    86   62   32   21   12  ...   
SPODNIE         1242   467   264   195   124   108   73   57   45   25  ...   
All            14077  4399  2594  1815  1338  1055  971  811  738  587  ...   

              Tożsamość odzieży                                   
Ilość prań                   23  24  25 26 27 28 32 35 36    All  
Opis                                                              
BLUZA GRANAT                  1   1   1  1  0  0  0  0  0   6679  
KAMIZELKA OC                  0   0   0  0  0  0  0  0  0   2010  
KURTKA OC.                    0   0   0  0  0  0  0  0  0   3687  
OGRODNICZ OC.                 0   0   0  0  0  0  0  0  0   2022  
OGRODNICZKI                  19  14   9  3  3  0  1  0  0   4013  
POLO                         11   4   4  3  0  3  0  1  1   6696  
SOFTSHELL                     0   0   0  0  0  0  0  0  0   3694  
SPODNIE                       0   1   2  0  0  1  0  0  0   2716  
All                          31  20  16  7  3  4  1  1  1  31517  

[9 rows x 198 columns]

#### Wczytujemy informacje o grupach przypisania

In [67]:
grupy = pd.read_csv(r'F:\grupy.csv', delimiter=';') # I'm sorry but I don't share data files

In [69]:
grupy.head()

Imię  Grupa
0  P0000751      2
1  C0000765      1
2  P0000067      3
3  P0000758      3
4  P0000691      3

In [70]:
df = df.merge(grupy, how='left', on='Imię')

In [71]:
df.head()

Nazwa Tożsamość odzieży          Opis  Ilość prań      Imię  \
0  POZNAŃ GROBLA       40000FA2717  BLUZA GRANAT           1  P0000751   
1  POZNAŃ GROBLA       40000FA2FD0  BLUZA GRANAT           0  C0000765   
2  POZNAŃ GROBLA       40000FA32B7  BLUZA GRANAT           0  C0000765   
3  POZNAŃ GROBLA       40000FA32ED  BLUZA GRANAT           0  P0000067   
4  POZNAŃ GROBLA       6C0001B13AC  BLUZA GRANAT           1  P0000758   

   Liczba prań dla aktualnego użytkownika odzieży.  Odział Jednostka  Grupa  
0                                                0  POZNAŃ    GROBLA      2  
1                                                0  POZNAŃ    GROBLA      1  
2                                                0  POZNAŃ    GROBLA      1  
3                                                0  POZNAŃ    GROBLA      3  
4                                                1  POZNAŃ    GROBLA      3

#### Zapisujemy data frame do pliku z znacznikiem daty

In [72]:
date = datetime.date.today()

In [73]:
print('database_{}'.format(date))

database_2023-02-10


In [74]:
excelWriter = pd.ExcelWriter(r'F:\test_{}.xlsx'.format(date))

In [75]:
df.to_excel(excelWriter)

In [76]:
excelWriter.save()

Tabele z znacznikiem daty są niezbędne do porównań zmian w ilościach prania na poszczególnych elementach wyposażenia. Co z kolei pozwoli na sprawdzenie z fakturą. Usługa prania jest fakturowana w cyklu tygodniowym. 
***
Wykorzystanie grupby po Oddziale a następnie funkcji get_group pozwala na przygotowanie automatycznych raportów dla pracowników danego oddziału